This notebook contains YOLOv8 training on a custom dataset with potholes.

All the models (nano, small, medium, etc) can be trained using the same procedures.

The dataset contains more then 7K images from videos with manual annotations.

dataset link = https://www.dropbox.com/s/qvglw8pqo16769f/pothole_dataset_v8.zip?dl=1

In [1]:
import ultralytics
# checking GPU access:
ultralytics.checks()

Ultralytics YOLOv8.0.230 🚀 Python-3.8.18 torch-2.1.2 CUDA:0 (NVIDIA GeForce GTX 1660, 6144MiB)
Setup complete ✅ (12 CPUs, 15.9 GB RAM, 553.7/930.9 GB disk)


Batch size was set to 8 to increase precision, but it also increases training time.

Additionally, imgsz (image resolution) was set to 1280 because refering to the dataset some pots are too small. Again, this increses training time.

Model will train for 10 epochs to reduce the training time because the dataset is big.

In [4]:
from ultralytics import YOLO
 
# Load the model.
model = YOLO('yolov8n.pt')
 
# Training.
results = model.train(
   data='datasets\pothole_dataset_v8\pothole_v8.yaml',
   imgsz=1280,
   epochs=10,
   batch=8,
   name='yolov8n_v8_10e'
)

New https://pypi.org/project/ultralytics/8.0.232 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.0.230 🚀 Python-3.8.18 torch-2.1.2 CUDA:0 (NVIDIA GeForce GTX 1660, 6144MiB)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=datasets\pothole_dataset_v8\pothole_v8.yaml, epochs=10, time=None, patience=50, batch=8, imgsz=1280, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=yolov8n_v8_50e, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show

train: Scanning C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\labels.cache... 6960 images, 10 backgrounds, 0 corrupt: 100%|██████████| 6962/6962 [00:00<?, ?it/s]

train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0012012.jpg: 2 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0052120.jpg: 1 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0063926.jpg: 1 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0063927.jpg: 1 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0064103.jpg: 2 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0064146.jpg: 1 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\train\images\G0064186.jpg: 1 duplicate labels removed
train: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v


val: Scanning C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\labels.cache... 271 images, 0 backgrounds, 0 corrupt: 100%|██████████| 271/271 [00:00<?, ?it/s]

val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011603.jpg: 1 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011614.jpg: 1 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011616.jpg: 2 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011661.jpg: 1 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011662.jpg: 1 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011664.jpg: 1 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G0011677.jpg: 1 duplicate labels removed
val: WARNING ⚠️ C:\Users\slyus\Desktop\conda_tf\datasets\pothole_dataset_v8\valid\images\G

Plotting labels to runs\detect\yolov8n_v8_50e\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
10 epochs...
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/871 [00:00<?, ?it/s]

ClearML Monitor: Could not detect iteration reporting, falling back to iterations as seconds-from-start


       1/10      7.15G      2.287       5.59      1.989         27       1280:  81%|████████  | 707/871 [1:18:15<18:09,  6.64s/it]


KeyboardInterrupt: 

Training may take up to several hours depending on your GPU (approximately 1:40 for 1 epoch using GPU GTX 1660).

Plots were saved to the local directory. It can be seen from the graphs that 10 epochs is not enough and the model can train more.

To evalutate the model:

In [ ]:
!yolo task=detect mode=val model=runs/detect/yolov8n_v8_50e/weights/best.pt name=yolov8n_eval data=pothole_v8.yaml imgsz=1280   # evaluting the best model

First, we used 'nano' model. We can also compare the results to s and bigger models.

In [ ]:
# Now we use S model
model_s = YOLO('yolov8s.pt')
results = model_s.train(
   data='datasets\pothole_dataset_v8\pothole_v8.yaml',
   imgsz=1280,
   epochs=10,
   batch=8,
   name='yolov8s_v8_10e'
)

In [ ]:
!yolo task=detect mode=val model=runs/detect/yolov8s_v8_50e/weights/best.pt name=yolov8s_eval data=pothole_v8.yaml imgsz=1280   # evaluting the best model

In [ ]:

model_m = YOLO('yolov8m.pt')
results = model_m.train(
   data='datasets\pothole_dataset_v8\pothole_v8.yaml',
   imgsz=1280,
   epochs=10,
   batch=8,
   name='yolov8m_v8_10e'
)

In [ ]:
!yolo task=detect mode=val model=runs/detect/yolov8m_v8_50e/weights/best.pt name=yolov8m_eval data=pothole_v8.yaml imgsz=1280   # evaluting the best model

To check how the models work and predict we can run:

In [ ]:
# download a video to the directory, for instance https://learnopencv.com/wp-content/uploads/2023/01/train-yolov8-custom-data-nano-result-1.mp4
!yolo task=detect mode=predict model=runs/detect/yolov8n_v8_50e/weights/best.pt source=inference_data/video.mp4 show=True imgsz=1280 name=yolov8n_v8_50e_infer1280 hide_labels=True

In [ ]:
!yolo task=detect mode=predict model=runs/detect/yolov8s_v8_50e/weights/best.pt source=inference_data/video.mp4 show=True imgsz=1280 name=yolov8s_v8_50e_infer1280 hide_labels=True

In [ ]:
!yolo task=detect mode=predict model=runs/detect/yolov8m_v8_50e/weights/best.pt source=inference_data/video.mp4 show=True imgsz=1280 name=yolov8m_v8_50e_infer1280 hide_labels=True

### Summary
All the models are pretty accurate, but none of them are flawless. Specifically, model n and m don't really stand out from each other. However, they seem to have the potential to get even better with more training sessions. At the moment, the small gap between them doesn't seem like a smart trade-off considering the significant gap in how well they predict real data (as N and S can be over 5 times slower in performance). The precision gap between the models might become clearer if we significantly increase the number of epochs them more time to train.

Materials taken from https://learnopencv.com/train-yolov8-on-custom-dataset/
The article also contain comparison of models' mAP. However, models there does not show big difference even using 50 epochs.